scenario.py

```diff
@@ -189,10 +190,216 @@ def __init__(self, params, experiment_path):
+
+        for i in range(self.partners_count):
+            partner = Partner(i)
+            self.partners_list.append(partner)
+
+    def split_data_advanced(self):
+        """Advanced split: Populates the partners with their train and test data (not pre-processed)"""
+
+        x_train = self.x_train
+        y_train = self.y_train
+        x_test = self.x_test
+        y_test = self.y_test
+        partners_list = self.partners_list
+        amounts_per_partner = self.amounts_per_partner
+        advanced_split_option = self.samples_split_option
+
+        # Compose the lists of partners with data samples from shared clusters and those with specific clusters
+        for p in partners_list:
+            p.cluster_count = advanced_split_option[p.id][0]
+            p.cluster_split_option = advanced_split_option[p.id][1]
+
+        partners_with_shared_clusters = [p for p in partners_list if p.cluster_split_option == 'shared']
+        partners_with_specific_clusters = [p for p in partners_list if p.cluster_split_option == 'specific']
+
+        # Compose the list of different labels in the dataset
+        labels = list(set(y_train))
+        random.seed(42)
+        random.shuffle(labels)
+
+        # Check coherence of the split option:
+        nb_diff_labels = len(labels)
+        sum_specific_clusters = sum([p.cluster_count for p in partners_with_specific_clusters])
+        if not partners_with_shared_clusters:
+            max_shared_clusters = 0
+        else:
+            max_shared_clusters = max([p.cluster_count for p in partners_with_shared_clusters])
+        assert sum_specific_clusters + max_shared_clusters  <= nb_diff_labels
+
+        # Stratify the dataset per labels
+        x_train_for_cluster, y_train_for_cluster, count_available_per_cluster = {}, {}, {}
+        for label in labels:
+            idx_in_full_trainset = np.where(y_train == label)
+            x_train_for_cluster[label] = x_train[idx_in_full_trainset]
+            y_train_for_cluster[label] = y_train[idx_in_full_trainset]
+            count_available_per_cluster[label] = len(y_train_for_cluster[label])
+
+        # Sort partners per nb of different clusters in descending order
+        partners_with_shared_clusters = sorted(
+            partners_with_shared_clusters,
+            key=operator.attrgetter("cluster_count"),
+            reverse=True,
+        )
+        partners_with_specific_clusters = sorted(
+            partners_with_specific_clusters,
+            key=operator.attrgetter("cluster_count"),
+            reverse=True,
+        )
+
+        # Initialize dict for storing which partners pick from which clusters
+        dict_specific, dict_shared = {}, {}
+
+        # First, partners getting data samples from clusters they don't share
+        position = 0
+        for p in partners_with_specific_clusters:
+            dict_specific[p.id] = {"clusters_list": labels[position:position + p.cluster_count]}
```

Comment:
It's a dict inside a dict ???

Reply:
So `clusters_list`contains the label for every parter? I would suggest renaming it `label_list`

Reply:
> It's a dict inside a dict ???

Yes, for each partner we init a dict which will be populated with the different temporary calculations (nb_available_samples, nb_samples_configured, nb_samples_configured_resized, nb_samples_configured_resized_per_cluster, final_nb_samples, final_nb_samples_p_cluster)

> So `clusters_list`contains the label for every parter? I would suggest renaming it `label_list`

I think `clusters` here is closer in terms of meaning. What is shared are the clusters of data samples, not the labels. The labels are the names of the clusters and the criteria that defines them.